lib/ClangImporter/ImportType.cpp

```diff
@@ -1156,208 +1195,246 @@ static bool isNSString(Type type) {
 }
 
-static ImportedType adjustTypeForConcreteImport(
-    ClangImporter::Implementation &impl,
-    ImportResult importResult, ImportTypeKind importKind,
-    bool allowNSUIntegerAsInt, Bridgeability bridging, OptionalTypeKind optKind,
-    bool resugarNSErrorPointer) {
-  Type importedType = importResult.AbstractType;
-  ImportHint hint = importResult.Hint;
-
-  if (importKind == ImportTypeKind::Abstract) {
-    return {importedType, false};
-  }
-
-  // 'void' can only be imported as a function result type.
-  if (hint == ImportHint::Void &&
-      (importKind == ImportTypeKind::AuditedResult ||
-       importKind == ImportTypeKind::Result)) {
-    return {impl.getNamedSwiftType(impl.getStdlibModule(), "Void"), false};
-  }
-
-  // Import NSString * globals as String.
-  if (hint == ImportHint::ObjCBridged && isNSString(importedType) &&
-      (importKind == ImportTypeKind::Variable ||
-       importKind == ImportTypeKind::AuditedVariable)) {
-    return {hint.BridgedType, false};
-  }
+static Type maybeImportNSErrorOutParameter(ClangImporter::Implementation &impl,
+                                           Type importedType,
+                                           bool resugarNSErrorPointer) {
+  PointerTypeKind PTK;
+  auto elementType = importedType->getAnyPointerElementType(PTK);
+  if (!elementType || PTK != PTK_AutoreleasingUnsafeMutablePointer)
+    return Type();
 
-  // For anything else, if we completely failed to import the type
-  // abstractly, give up now.
-  if (!importedType)
-    return {Type(), false};
+  auto elementObj = elementType->getOptionalObjectType();
+  if (!elementObj)
+    return Type();
 
-  // Special case AutoreleasingUnsafeMutablePointer<NSError?> parameters.
-  auto maybeImportNSErrorPointer = [&]() -> Type {
-    if (importKind != ImportTypeKind::Parameter)
-      return Type();
+  auto elementClass = elementObj->getClassOrBoundGenericClass();
+  if (!elementClass)
+    return Type();
 
-    PointerTypeKind PTK;
-    auto elementType = importedType->getAnyPointerElementType(PTK);
-    if (!elementType || PTK != PTK_AutoreleasingUnsafeMutablePointer)
-      return Type();
+  // FIXME: Avoid string comparison by caching this identifier.
+  if (elementClass->getName().str() !=
+        impl.SwiftContext.getSwiftName(KnownFoundationEntity::NSError))
+    return Type();
 
-    auto elementObj = elementType->getOptionalObjectType();
-    if (!elementObj)
-      return Type();
+  ModuleDecl *foundationModule = impl.tryLoadFoundationModule();
+  if (!foundationModule ||
+      foundationModule->getName()
+        != elementClass->getModuleContext()->getName())
+    return Type();
 
-    auto elementClass = elementObj->getClassOrBoundGenericClass();
-    if (!elementClass)
-      return Type();
 
-    // FIXME: Avoid string comparison by caching this identifier.
-    if (elementClass->getName().str() !=
-          impl.SwiftContext.getSwiftName(KnownFoundationEntity::NSError))
-      return Type();
```

Comment:
Ah, I see, this code was already here and might have predated `getSwiftId()`.

Come to think of it, I think there might be an `ASTContext::getNSErrorDecl()`. Can you use that to abridge some of this logic?

Reply:
For now I think I'd like to leave it as is; the code can always be improved further in the future.